hazelcast/src/hazelcast/client/sql.cpp

```diff
@@ -238,6 +314,55 @@ sql_service::query_connection()
     return connection;
 }
 
+std::shared_ptr<connection::Connection>
+sql_service::query_connection(int32_t partition_id)
+{
+    std::shared_ptr<connection::Connection> connection;
+    try {
+        const auto node_id =
+          client_context_.get_partition_service().get_partition_owner(
+            partition_id);
+
+        if (node_id.is_nil()) {
+            return query_connection();
+        }
+
+        connection =
+          client_context_.get_connection_manager().get_connection(node_id);
+
+        if (connection == nullptr) {
+            return query_connection();
+        }
+    } catch (const std::exception& e) {
+        rethrow(e);
+    }
+
+    return connection;
+}
+
+boost::optional<int32_t>
+sql_service::extract_partition_id(const sql_statement& statement,
+                                  int32_t arg_index) const
+{
+    if (!is_smart_routing_) {
+        return boost::none;
+    }
+
+    if (statement.serialized_parameters_.size() == 0) {
+        return boost::none;
+    }
+
+    if (arg_index >=
+          static_cast<int32_t>(statement.serialized_parameters_.size()) ||
+        arg_index < 0) {
+        return boost::none;
+    }
+
+    const auto key = statement.serialized_parameters_[arg_index];
```

Comment:
value copy for the whole data binary? why not 
     const auto &key = statement.serialized_parameters_[arg_index];

Reply:
Fixed. As this PR is merged, the changes are made in another branch. https://github.com/hazelcast/hazelcast-cpp-client/pull/1188